### PR TITLE
Add OOO  request approved  data to user future status

### PR DESCRIPTION
**Pull request by sahsisunny**
Date: 06 Apr 2024
Developer Name: @sahsisunny 

## Issue Ticket Number
- #1998 

## Description

This pull request addresses the issue where the user's status was not being updated after the approval of an Out of Office (OOO) request. The system has been updated to ensure that after an OOO request is approved, the user's status is automatically changed to either their normal status or OOO status based on the system's configuration.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

Local testing during development confirms the functionality works as expected.

## Screenshots
<details>
<summary>Demo Video</summary>

https://github.com/Real-Dev-Squad/website-backend/assets/70854507/2ec9c652-2020-4e9f-98ee-6409b84a0632

</details>

## Test Coverage
<details>
<summary>Test Coverage</summary>

| File                   | Statements | Branches | Functions | Lines | Uncovered Lines |
|------------------------|------------|----------|-----------|-------|-----------------|
| models/userFutureStatus.ts | 86.2     | 50       | 100       | 84    | 23-24,53-54     |
| models/userStatus.js       | 59.51    | 46.52    | 72.22     | 58.45 | 1-492,570,587-588,609-611,641-642,658-659,685-689,706-707 |
| models/requests.ts      | 86.74      | 89.74    | 100       | 85.71 | 41-42,76-77,115,124-125,127-128,162-163 |

</details>

## Additional Notes

No additional notes.

### CODE REVIEW

File: models/userStatus.js [hunk 1]
```diff
@@ -673,6 +674,40 @@ const cancelOooStatus = async (userId) => {
...
+    const userStatusDocs = await userStatusModel.where("userId", "==", futureStatusData.userId).limit(1).get();
+    const [userStatusDoc] = userStatusDocs.docs;
+    let docId;
+    let userStatusData;
+
+    if (userStatusDoc) {
+      docId = userStatusDoc.id;
+      userStatusData = userStatusDoc.data();
+
+      if (userStatusData.futureStatus && userStatusData.futureStatus.state === "approved") {
```
Comment (palakgupta2712) [inline models/userStatus.js:688]: can you please use constant here?

**Review by palakgupta2712 — CHANGES_REQUESTED** [review]

File: models/userStatus.js [hunk 2]
```diff
@@ -673,6 +674,40 @@ const cancelOooStatus = async (userId) => {
...
+        return { id: docId, userStatusExists: true, data: userStatusData };
+      }
+    } else {
+      const newUserStatusRef = userStatusModel.doc();
+      await newUserStatusRef.set({ userId: futureStatusData.userId });
+      docId = newUserStatusRef.id;
+      userStatusData = { userId: futureStatusData.userId };
+    }
+
+    delete futureStatusData.userId;
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply models/userStatus.js:698]: which delete operator?

**Review by Achintya-Chatterjee — APPROVED** [review]

File: models/userStatus.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (palakgupta2712) [reply models/userStatus.js:694]: delete futureStatusData.userId;

[the remark this answers is not in the document]
Comment (sahsisunny) [reply models/userStatus.js:694]: I am using delete operator here, which delete operator should I use?

File: models/userFutureStatus.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+const userFutureStatusModel = firestore.collection("userFutureStatus");
+import { UserFutureStatusType } from "../types/userFutureStatus";
+import * as admin from "firebase-admin";
+
+/**
+ * Function to create user future status
+ * @param body: UserFutureStatusType
+ * @returns UserFutureStatusType
+ */
+export const createUserFutureStatus = async (body: UserFutureStatusType) => {
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply models/userFutureStatus.ts:11]: It's suggested by @lakshayman

**Review by palakgupta2712 — APPROVED** [review]

**Pull request merged**